src/PackageManagement/IDE/DeleteOnRestartManager.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.PackageManagement
+{
+    public sealed class DeleteOnRestartManager : IDeleteOnRestartManager
+    {
+        // The file extension to add to the empty files which will be placed adjacent to partially uninstalled package
+        // directories marking them for removal the next time the solution is opened.
+        private const string DeletionMarkerSuffix = ".deleteme";
+        private const string DeletionMarkerFilter = "*" + DeletionMarkerSuffix;
+
+        private readonly FolderNuGetProject _folderNugetProject;
+
+        private ISolutionManager SolutionManager { get; }
+
+        private ISettings Settings { get; }
+
+        /// <summary>
+        /// Create a new instance of <see cref="DeleteOnRestartManager"/>.
+        /// </summary>
+        /// <param name="project"></param>
+        public DeleteOnRestartManager(NuGetProject project)
+        {
+            _folderNugetProject = project as FolderNuGetProject;
+            if (_folderNugetProject == null)
+            {
+                _folderNugetProject = (project as MSBuildNuGetProject)?.FolderNuGetProject;
+            }
+
+            if (_folderNugetProject == null)
+            {
+                // TODO: should this throw? Replace with a proper message.
```

Comment:
you want to add an argument null exception on top as well, and add a reasource string here